client/protocol.py

```diff
@@ -0,0 +1,200 @@
+
+    @classmethod
+    def build_from_addrs(cls, addrs: list[str]) -> ProtocolClient:
+        """Build client from addresses, this method will fetch all members from servers"""
+
+        stubs: list[ProtocolStub] = []
+
+        for addr in addrs:
+            channel = grpc.insecure_channel(addr)
+            stub = ProtocolStub(channel)
+            stubs.append(stub)
+
+        cluster = fetch_cluster(stubs)
+
+        return cls(
+            cluster.leader_id,
+            stubs,
+            cluster.members,
+        )
+
+    def propose(self, cmd: Command, use_fast_path: bool = False) -> tuple[CommandResponse, SyncResponse | None]:
+        """Propose the request to servers, if use_fast_path is false, it will wait for the synced index"""
+
+        if use_fast_path:
+            try:
+                cmd_res = asyncio.run(self.fast_round(cmd))
+                return (cmd_res, None)
+            except Exception:
+                try:
+                    sync_res, cmd_res = asyncio.run(self.slow_round(cmd))
+                    return (cmd_res, sync_res)
+                except Exception as e:
+                    raise e
+        else:
+            asyncio.run(self.fast_round(cmd))
+            try:
+                sync_res, cmd_res = asyncio.run(self.slow_round(cmd))
+                return (cmd_res, sync_res)
+            except Exception as e:
+                raise e
+
+    async def fast_round(self, cmd: Command) -> CommandResponse:
+        """
+        The fast round of Curp protocol
+        It broadcast the requests to all the curp servers.
+        """
+
+        logging.info("fast round start. propose id: %s", cmd.propose_id)
+
+        ok_cnt = 0
+        is_received_leader_res = False
+        is_received_success_res = None
+        cmd_res = CommandResponse()
+        exe_err = ExecuteError(_ExecuteError())
+        propose_err = ProposeError(_ProposeError())
+
+        for stub in self.inner:
+            res = await propose_wrapper(stub, cmd)
+
+            if res.HasField("result"):
+                cmd_result = res.result
+                ok_cnt += 1
+                is_received_leader_res = True
+                if cmd_result.HasField("er"):
+                    is_received_success_res = True
+                    cmd_res.ParseFromString(cmd_result.er)
+                if cmd_result.HasField("error"):
+                    is_received_success_res = False
+                    exe_err.inner.ParseFromString(cmd_result.error)
+            elif res.HasField("error"):
+                is_received_leader_res = True
+                propose_err.inner.CopyFrom(res.error)
+            else:
+                ok_cnt += 1
+
+            if is_received_leader_res and ok_cnt >= super_quorum(len(self.connects)):
+                logging.info("fast round success")
+                if is_received_success_res:
+                    return cmd_res
+                if not is_received_success_res:
+                    raise exe_err
```

Comment:
Here it looks like sending and processing the propose one by one, we should send them all at once

Reply:
`propose_wrapper` is an async function that doesn't block the main function.

Reply:
But await `propose_wrapper` will block the for loop until `propose_wrapper` return

Reply:
You're right. I'll make the changes you suggested.